packages/medusa/src/services/payment-provider.ts

```diff
@@ -165,55 +174,59 @@ export default class PaymentProviderService extends TransactionBaseService {
-   * @param providerId - the id of the provider to create payment with
+   * @param providerIdOrSessionInput - the id of the provider to create payment with or the input data
    * @param cart - a cart object used to calculate the amount, etc. from
    * @return the payment session
    */
-  async createSession(providerId: string, cart: Cart): Promise<PaymentSession> {
+  async createSession<
+    TInput extends string | PaymentSessionInput = string | PaymentSessionInput
+  >(
+    providerIdOrSessionInput: TInput,
+    ...[cart]: TInput extends string ? [Cart] : [never?]
+  ): Promise<PaymentSession> {
     return await this.atomicPhase_(async (transactionManager) => {
-      const provider = this.retrieveProvider(providerId)
-      const sessionData = await provider
-        .withTransaction(transactionManager)
-        .createPayment(cart)
+      const providerId = isString(providerIdOrSessionInput)
+        ? providerIdOrSessionInput
+        : providerIdOrSessionInput.provider_id
+      const data = (
+        isString(providerIdOrSessionInput) ? cart : providerIdOrSessionInput
+      ) as Cart | PaymentSessionInput
 
-      const sessionRepo = transactionManager.getCustomRepository(
-        this.paymentSessionRepository_
-      )
+      const provider = this.retrieveProvider<AbstractPaymentService>(providerId)
+      const context = this.buildPaymentContext(data)
 
-      const toCreate = {
-        cart_id: cart.id,
-        provider_id: providerId,
-        data: sessionData,
-        status: "pending",
+      if (!isDefined(context.currency_code) || !isDefined(context.amount)) {
+        throw new MedusaError(
+          MedusaError.Types.INVALID_ARGUMENT,
+          "`currency_code` and `amount` are required to create payment session."
+        )
       }
 
-      const created = sessionRepo.create(toCreate)
-      return await sessionRepo.save(created)
-    })
-  }
-
-  async createSessionNew(
-    sessionInput: PaymentProviderDataInput
-  ): Promise<PaymentSession> {
-    return await this.atomicPhase_(async (transactionManager) => {
-      const provider = this.retrieveProvider(sessionInput.provider_id)
-      const sessionData = await provider
+      const paymentResponse = await provider
         .withTransaction(transactionManager)
-        .createPaymentNew(sessionInput)
+        .createPayment(context)
 
-      const sessionRepo = transactionManager.getCustomRepository(
-        this.paymentSessionRepository_
+      const sessionData = paymentResponse.session_data ?? paymentResponse
+
+      await this.processUpdateRequestsData(
+        {
+          customer: { id: context.customer?.id },
+        },
+        paymentResponse
       )
 
-      const toCreate = {
-        provider_id: sessionInput.provider_id,
-        data: sessionData,
-        status: "pending",
-        amount: sessionInput.amount,
-      } as PaymentSession
+      const amount = this.featureFlagRouter_.isFeatureEnabled(
```

Comment:
suggestion: Not entirely sure why this is necessary. Can I get you to add a comment for when this needs revisiting in the future?

Reply:
Since we decided to merge the new method in that pr, it must take into account the payment collection. The amount on the payment session does not exists unless the order edit flag is enabled